federation/pkg/federation-controller/util/delaying_deliverer.go

```diff
@@ -0,0 +1,144 @@
+	"container/heap"
+	"time"
+)
+
+const (
+	delayingDelivererUpdateChanCapacity = 1000
+)
+
+// DelayingDelivererItem is structure delivered by DelayingDeliverer to the
+// target channel.
+type DelayingDelivererItem struct {
+	// Key under which the value was added to deliverer.
+	Key string
+	// Value of the item.
+	Value interface{}
+	// When the item was enqueued in the deliverer.
+	EnqueueTime time.Time
+	// When the item should be delivered.
+	DeliveryTime time.Time
+}
+
+type delivererHeap struct {
+	keyPosition map[string]int
+	data        []*DelayingDelivererItem
+}
+
+// Functions required by container.Heap.
+
+func (dh *delivererHeap) Len() int { return len(dh.data) }
+func (dh *delivererHeap) Less(i, j int) bool {
+	return dh.data[i].DeliveryTime.Before(dh.data[j].DeliveryTime)
+}
+func (dh *delivererHeap) Swap(i, j int) {
+	dh.keyPosition[dh.data[i].Key] = j
+	dh.keyPosition[dh.data[j].Key] = i
+	dh.data[i], dh.data[j] = dh.data[j], dh.data[i]
+}
+
+func (dh *delivererHeap) Push(x interface{}) {
+	item := x.(*DelayingDelivererItem)
+	dh.data = append(dh.data, item)
+	dh.keyPosition[item.Key] = len(dh.data) - 1
+}
+
+func (dh *delivererHeap) Pop() interface{} {
+	n := len(dh.data)
+	item := dh.data[n-1]
+	dh.data = dh.data[0 : n-1]
+	delete(dh.keyPosition, item.Key)
+	return item
+}
+
+// A structure that pushes the items to the target channel at a given time.
+type DelayingDeliverer struct {
+	targetChannel chan *DelayingDelivererItem
+
+	heap          *delivererHeap
+	updateChannel chan *DelayingDelivererItem
+	stopChannel   chan struct{}
+}
+
+func NewDelayingDeliverer(targetChannel chan *DelayingDelivererItem) *DelayingDeliverer {
+	d := &DelayingDeliverer{
+		targetChannel: targetChannel,
+		heap: &delivererHeap{
+			keyPosition: make(map[string]int),
+			data:        make([]*DelayingDelivererItem, 0),
+		},
+		updateChannel: make(chan *DelayingDelivererItem, delayingDelivererUpdateChanCapacity),
+		stopChannel:   make(chan struct{}),
+	}
+	go d.run()
+	return d
+}
+
+func (d *DelayingDeliverer) run() {
+	for {
+		now := time.Now()
+
+	deliveryLoop:
+		for d.heap.Len() > 0 {
+			if now.Before(d.heap.data[0].DeliveryTime) {
+				break deliveryLoop
+			}
+			item := heap.Pop(d.heap).(*DelayingDelivererItem)
+			d.targetChannel <- item
+		}
+		nextWakeUp := now.Add(time.Hour)
+		if d.heap.Len() > 0 {
+			nextWakeUp = d.heap.data[0].DeliveryTime
+		}
+		sleepTime := nextWakeUp.Sub(now)
+
+	main_select:
+		select {
+		case <-time.After(sleepTime):
+			break main_select // just wake up and process the data
+		case item := <-d.updateChannel:
+			if position, found := d.heap.keyPosition[item.Key]; found {
+				if item.DeliveryTime.Before(d.heap.data[position].DeliveryTime) {
+					d.heap.data[position] = item
+					heap.Fix(d.heap, position)
+				}
+				// Ignore if later.
+			} else {
+				heap.Push(d.heap, item)
+			}
+		case <-d.stopChannel:
+			return
+		}
+	}
+}
+
+func (d *DelayingDeliverer) Stop() {
+	d.stopChannel <- struct{}{}
```

Comment:
I meant only close(d.stopChannel). You don't need to send anything:

close(d.stopChannel) will trigger <-d.stoppChannel too.